### ISSUE TITLE
exclusiveContent API does not work with Kotlin Gradle DSL

### ISSUE THREAD

**Issue by ZacSweers** [gradle/gradle#12452]
### Expected Behavior 
It does not crash

### Current Behavior
It crashes with this message:

```
Build file '/Users/zsweers/dev/android/personal/CatchUp/build.gradle.kts' line: 34

Cannot mutate content repository descriptor after repository has been used
```

for this snippet in kotlin gradle DSL

```kotlin
buildscript {
  repositories {
    exclusiveContent {
      forRepository {
        // For R8/D8 releases
        maven("https://storage.googleapis.com/r8-releases/raw")
      }
      filter {
        includeModule("com.android.tools", "r8")
      }
    }
  }
}
```

### Context
This API works fine from groovy but does not work when applied via kotlin gradle DSL

### Steps to Reproduce 
Here is a repro branch/project: https://github.com/ZacSweers/CatchUp/tree/z/exlusiveContentBug

Demo commit: https://github.com/ZacSweers/CatchUp/commit/fe2bec6d3bd36f5131b93aa704da5e8f8a42ac08

### Your Environment
Cannot perform a build scan due to failure

**Comment by bmuskalla** [comment]
Reproducer

```kotlin
buildscript {
  repositories {
    mavenCentral()
    exclusiveContent {
      forRepository {
        maven("https://storage.googleapis.com/r8-releases/raw") {
          metadataSources {
            artifact()
          }
        }
      }
      filter {
        includeModule("com.android.tools", "r8")
      }
    }
  }
}

plugins {
  id("com.github.plnice.canidropjetifier") version "0.5"
}
```

**Comment by ZacSweers** [comment]
This is still broken in 6.4

```
Caused by: java.lang.IllegalStateException: Cannot mutate content repository descriptor 'maven(https://maven-central.storage-download.googleapis.com/repos/central/data/)' after repository has been used
	at org.gradle.api.internal.artifacts.repositories.DefaultRepositoryContentDescriptor.assertMutable(DefaultRepositoryContentDescriptor.java:52)
	at org.gradle.api.internal.artifacts.repositories.DefaultRepositoryContentDescriptor.addExclude(DefaultRepositoryContentDescriptor.java:187)
	at org.gradle.api.internal.artifacts.repositories.DefaultRepositoryContentDescriptor.excludeModule(DefaultRepositoryContentDescriptor.java:160)
	at org.gradle.api.internal.artifacts.dsl.DefaultRepositoryHandler$1.includeModule(DefaultRepositoryHandler.java:185)
	at slack.gradle.SlackPluginKt$configureSlackRootBuildscript$repositoryConfig$1$2$2.execute(SlackPlugin.kt:262)
	at slack.gradle.SlackPluginKt$configureSlackRootBuildscript$repositoryConfig$1$2$2.execute(SlackPlugin.kt)
	at org.gradle.api.internal.artifacts.dsl.DefaultRepositoryHandler.lambda$transformForExclusivity$0(DefaultRepositoryHandler.java:172)
	at org.gradle.api.internal.artifacts.repositories.AbstractArtifactRepository.content(AbstractArtifactRepository.java:123)
	at org.gradle.api.internal.artifacts.dsl.DefaultRepositoryHandler$ExclusiveContentRepositorySpec.lambda$apply$2(DefaultRepositoryHandler.java:247)
	at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction$1$1.run(DefaultCollectionCallbackActionDecorator.java:100)
	at org.gradle.configuration.internal.DefaultUserCodeApplicationContext.reapply(DefaultUserCodeApplicationContext.java:60)
	at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction$1.run(DefaultCollectionCallbackActionDecorator.java:97)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
	at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction.execute(DefaultCollectionCallbackActionDecorator.java:94)
	at org.gradle.api.internal.DefaultDomainObjectCollection.all(DefaultDomainObjectCollection.java:163)
	at org.gradle.api.internal.artifacts.dsl.DefaultRepositoryHandler$ExclusiveContentRepositorySpec.apply(DefaultRepositoryHandler.java:243)
	at org.gradle.api.internal.artifacts.dsl.DefaultRepositoryHandler.exclusiveContent(DefaultRepositoryHandler.java:167)
	at slack.gradle.SlackPluginKt$configureSlackRootBuildscript$repositoryConfig$1.invoke(SlackPlugin.kt:256)
	at slack.gradle.SlackPluginKt$configureSlackRootBuildscript$repositoryConfig$1.invoke(SlackPlugin.kt)
	at org.gradle.kotlin.dsl.ProjectExtensionsKt.repositories(ProjectExtensions.kt:154)
	at slack.gradle.SlackPluginKt$configureSlackRootBuildscript$1.invoke(SlackPlugin.kt:274)
	at slack.gradle.SlackPluginKt$configureSlackRootBuildscript$1.invoke(SlackPlugin.kt)
	at org.gradle.kotlin.dsl.ProjectExtensionsKt.buildscript(ProjectExtensions.kt:263)
	at slack.gradle.SlackPluginKt.configureSlackRootBuildscript(SlackPlugin.kt:273)
	at slack.gradle.SlackPluginKt.access$configureSlackRootBuildscript(SlackPlugin.kt:1)
	at slack.gradle.SlackPlugin.configureRootProject(SlackPlugin.kt:71)
	at slack.gradle.SlackPlugin.apply(SlackPlugin.kt:61)
	at slack.gradle.SlackPlugin.apply(SlackPlugin.kt:44)
	at org.gradle.api.internal.plugins.ImperativeOnlyPluginTarget.applyImperative(ImperativeOnlyPluginTarget.java:43)
	at org.gradle.api.internal.plugins.RuleBasedPluginTarget.applyImperative(RuleBasedPluginTarget.java:51)
	at org.gradle.api.internal.plugins.DefaultPluginManager.addPlugin(DefaultPluginManager.java:181)
	at org.gradle.api.internal.plugins.DefaultPluginManager.access$300(DefaultPluginManager.java:51)
	at org.gradle.api.internal.plugins.DefaultPluginManager$AddPluginBuildOperation.run(DefaultPluginManager.java:276)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
	at org.gradle.api.internal.plugins.DefaultPluginManager$2.execute(DefaultPluginManager.java:159)
	at org.gradle.api.internal.plugins.DefaultPluginManager$2.execute(DefaultPluginManager.java:156)
	at org.gradle.configuration.internal.DefaultUserCodeApplicationContext.apply(DefaultUserCodeApplicationContext.java:49)
	at org.gradle.api.internal.plugins.DefaultPluginManager.doApply(DefaultPluginManager.java:156)
```

**Comment by bmuskalla** [comment]
Keeping this closed as one usecase was fixed for 6.4. Opened #13038 to be fixed for 6.5
